### DISCUSSION TITLE
Why do GoQuorum writes block even without any transaction?

### DISCUSSION THREAD

**Discussion by gadginir** [ConsenSys/quorum#1416] [Q&A] [answered]
Hi,
I have observed that without any transaction, GoQuorum writes a block in the chain.
I can understand that this behavior is valid for Public chains where they want to increase the difficulty level for PoW consensus over the period of time to make it difficult to mine a block for Miners.
But why such feature is needed for private blockchain like GoQuorum? Is there any setting/configuration to change this behavior?

**Comment by baptiste-b-pegasys** [comment]
We are currently figuring out what to change in order to add a possible option for producing less empty blocks. There is no simple answer for multiple reasons.

We want to delay the empty block production, not stop them completely, because some smart contracts may rely on the 'now' instruction to have a recent timestamp on read calls.

In solidity 0.7, 'now' is deprecated, 'block.timestamp' should be used. There is no way to access a real 'now' like any OS as it relies on the latest block's timestamp.

---

There is an EIP for IBFT and an early implementation made. (https://media.consensys.net/scaling-consensus-for-enterprise-explaining-the-ibft-algorithm-ba86182ea668)

Compared to Raft in GoQuorum, which is a not a standard eth miner and doesn't produce empty blocks, IBFT relies on the eth miner process and structure, and also, to maintain the network (liveness), proposal has to be exchanged between miners on expected time in order to round change as expected.

It looks like there is a rule where the IBFT proposal ID is related to the block ID. (I read somewhere that the IBFT round number is included in the block)

https://pmg.csail.mit.edu/papers/osdi99.pdf

**Comment by gadginir** [comment]
Thanks for quick reply. I found` --istanbul.blockperiod` argument parameter. Does it work to delay the empty blocks writing on chain? 

1. Let's assume, if I keep its value 5, then does it write block only after 5 seconds? 
2. If there are enough transactions available with the allowed gas limit per block in 2 seconds, then is it still wait for 5 seconds or go ahead and write a block in 2 seconds?
3. If there are only few transactions available and it's not meeting a criteria of the gas limit, then is it going to wait for 5 sec to write a block?

**Comment by baptiste-b-pegasys** [comment, marked as answer]
It is 5 seconds period between blocks. (with or without tx)

Gas limit is related to the operations made by the contracts for executing transactions.
https://ethereum.org/en/developers/docs/blocks/#block-size

I understand that reaching this limit will end on pausing the execution of transactions until the block is released, therefore you are wasting computing time. (yet I have check it's the case in reality)

Anyway we propose 1 sec because it provides the best throughput.
